pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+                    msg = ('{name}.from_tuples requires tuples of '
+                           'length 2, got {tpl}').format(name=name, tpl=d)
+                    raise ValueError(msg)
+                except TypeError:
+                    msg = ('{name}.from_tuples received an invalid '
+                           'item, {tpl}').format(name=name, tpl=d)
+                    raise TypeError(msg)
+                lhs, rhs = d
+            left.append(lhs)
+            right.append(rhs)
+
+        return cls.from_arrays(left, right, closed, copy=False,
+                               dtype=dtype)
+
+    def _validate(self):
+        """Verify that the IntervalIndex is valid.
+
+        Checks that
+
+        * closed is valid
+        * left and right match lengths
+        * left and right have the same missing values
+        * left is always below right
+        """
+        if self.closed not in _VALID_CLOSED:
+            raise ValueError("invalid option for 'closed': {closed}"
+                             .format(closed=self.closed))
+        if len(self.left) != len(self.right):
+            raise ValueError('left and right must have the same length')
+        left_mask = notna(self.left)
+        right_mask = notna(self.right)
+        if not (left_mask == right_mask).all():
+            raise ValueError('missing values must be missing in the same '
+                             'location both left and right sides')
+        if not (self.left[left_mask] <= self.right[left_mask]).all():
+            raise ValueError('left side of interval must be <= right side')
+
+    # ---------
+    # Interface
+    # ---------
+    def __iter__(self):
+        return iter(self.values)
+
+    def __len__(self):
+        return len(self.left)
+
+    def __getitem__(self, value):
+        mask = self.isna()[value]
+        if is_scalar(mask) and mask:
+            return self._fill_value
+
+        left = self.left[value]
+        right = self.right[value]
+
+        # scalar
+        if not isinstance(left, Index):
+            return Interval(left, right, self.closed)
+
+        return self._shallow_copy(left, right)
+
+    def __setitem__(self, key, value):
+        if not (is_interval_dtype(value) or isinstance(value, Interval)):
+            msg = "'value' should be an interval type, got {} instead."
+            raise TypeError(msg.format(type(value)))
+
+        if value.closed != self.closed:
+            msg = "'value.closed' ({}) does not match {}."
+            raise ValueError(value.closed, self.closed)
+
+        # Need to ensure that left and right are updated atomically, so we're
+        # forced to copy, update the copy, and swap in the new values.
+        left = self.left.copy(deep=True)
+        right = self.right.copy(deep=True)
+
+        left.values[key] = value.left
+        right.values[key] = value.right
+
+        self._left = left
+        self._right = right
+
+    def fillna(self, value=None, method=None, limit=None):
```

Comment:
Should we do some extra checks on `value`?  Right now if I pass an `Interval` with a different `closed`, it silently gets changed to match the array's `closed`:

```python
In [2]: ia = pd.interval_range(0, 4).insert(0, np.nan).values

In [3]: ia
Out[3]:
IntervalArray([nan, (0.0, 1.0], (1.0, 2.0], (2.0, 3.0], (3.0, 4.0]],
              closed='right',
              dtype='interval[float64]')

In [4]: ia.fillna(pd.Interval(10, 20, closed='both'))
Out[4]:
IntervalArray([(10.0, 20.0], (0.0, 1.0], (1.0, 2.0], (2.0, 3.0], (3.0, 4.0]],
              closed='right',
              dtype='interval[float64]')
```
Could probably create some type of helper function/method for this, as I'd guess it'd be somewhat of a common thing to do.